### PR TITLE
Add interceptor to library

### PR DESCRIPTION
**Pull request by enyia21**

### CODE REVIEW

File: src/main/java/formflow/library/interceptors/DataRequiredInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import org.springframework.util.AntPathMatcher;
+import org.springframework.web.servlet.HandlerInterceptor;
+
+/**
+ * This interceptor prevents users from jumping to random pages in a flow.
+ */
+@Component
+@Slf4j
+@ConditionalOnProperty(name = "form-flow.session-continuity-interceptor.enabled", havingValue = "true")
+public class DataRequiredInterceptor implements HandlerInterceptor, Ordered {
```
Comment (bseeger) [inline src/main/java/formflow/library/interceptors/DataRequiredInterceptor.java:24]: For naming: I can see that because we require both session and submission id, that it's data that's required, but maybe this would be more aptly named `SessionContinuityInterceptor` to make it immediately recognizable (to programmers) that it's related to the configuration parameter?

File: README.md [hunk 2]
```diff
@@ -48,6 +48,11 @@ Table of Contents
     * [Sending Email](#sending-email)
         * [Mailgun](#mailgun)
     * [Localization](#localization)
+    * [Interceptors](#interceptors)
+        * [DataRequiredInterceptor](#datarequiredinterceptor)
+            * [Configuration](#configuration)
```
Comment (bseeger) [inline README.md:53]: Please only go three levels deep here, to keep the TOC shorter.   (keep the sections, just don't list them here beyond three levels. Thanks!)

File: README.md [hunk 3]
```diff
@@ -2182,6 +2187,85 @@ in [our JavaScript](https://github.com/codeforamerica/form-flow/blob/9a1c6f7a253
 
 We send the whole url to Mixpanel, including any query parameters, such as `lang`.
 
+## Interceptors
```
Comment (bseeger) [inline README.md:2190]: We need a separate section about landmark pages, independent of this section.  This would be where we explain what landmark pages are intended for and catalog the ones that exist (right now, only the firstScreen, but I anticipate more soon).

File: src/main/java/formflow/library/interceptors/DataRequiredInterceptor.java (hunk 1 above)
Comment (enyia21) [reply src/main/java/formflow/library/interceptors/DataRequiredInterceptor.java:24]: I like it!  I'll get that changed

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      throws Exception {
+    try {
+      String pathFormat = request.getRequestURI().contains("navigation") ? NAVIGATION_FLOW_PATH_FORMAT : FLOW_PATH_FORMAT;
+      Map<String, String> parsedUrl = new AntPathMatcher().extractUriTemplateVariables(pathFormat, request.getRequestURI());
+      String redirect_url = "/";
+      String screen = parsedUrl.get("screen");
+      String flow = parsedUrl.get("flow");
+      HttpSession session = request.getSession(false);
+      FlowConfiguration flowConfiguration = flowConfigurations.stream().filter(fc -> fc.getName().equals(flow)).findFirst()
+          .orElse(null);
```
Comment (mpidcock) [inline src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:54]: This block of variables is a bit dense. Some suggestions for what I am used to:
* `redirect_url` seems like a class level variable, or maybe something that should be configured as part of each landmark config
* `screen` and `session` could be moved further down in the code to right before they are used - clearer to read and ensures they are only declared if needed
   * I might also inline `screen` since it is only used once
* `flow` is also only used once, but the line is rather long already. Here is the way I normally make `stream()` lines more readable:
```
FlowConfiguration flowConfiguration = flowConfigurations.stream()
          .filter(fc -> fc.getName().equals(parsedUrl.get("flow")))
          .findFirst()
          .orElse(null);
``` 
basically every chained dot method on the stream gets its own line

**Review by mpidcock — APPROVED** [review]
My comments are all codestyle and might differ from team norms so I won't demand that you change anything, but look it over at least

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java (hunk 4 above)
Comment (bseeger) [reply src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:54]: I like to write streams this way as well, as I find it more readable.   I'll bring this to computer gardening to see if we can set a code style for this for the team.

**Comment by bseeger** [comment]
@mpidcock and @enyia21  - Thanks for reviewing, Martha.  I know Martha is around on Monday. Not sure of @enyia21 schedule. I think it's worth working through some of these comments and updating the code based on the feedback.   It is well worth cleaning up some of the things, like exception handling, to make it more clear what's going on here.

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  /**
+   * Sets the value of the interceptor to the highest integer value making it the last interceptor executed.
+   *
+   * @return Max Integer value.
+   */
+  @Override
+  public int getOrder() {
+    // Max value ensures that this interceptor is executed last.
```
[the remark this answers is not in the document]
Comment (enyia21) [reply src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:114]: ✔️ Removed the comment.   We should rely on the javadoc to explain this method

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+      if (session.getAttribute("id") == null) {
+        log.error("A submission ID was not found in the session for request to {}. Redirecting to landing page.",
+            request.getRequestURI());
+        response.sendRedirect(redirect_url);
+        return false;
+      }
+
+      return true;
+    } catch (IllegalStateException e) {
```
[the remark this answers is not in the document]
Comment (enyia21) [reply src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:102]: Yeah that feels weird as well.  I'm taking it out to see what breaks.   The prehandler still works the same in test

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      return true;
```
[the remark this answers is not in the document]
Comment (enyia21) [reply src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:103]: I removed the IllegalStateException for now.  I'm not sure if its necessary

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 8]
```diff
@@ -40,61 +41,63 @@ public SessionContinuityInterceptor(List<FlowConfiguration> flowConfigurations)
...
+        .orElse(null);
 
-      if (flowConfiguration == null) {
+    if (flowConfiguration == null) {
+      return true;
+    }
+
+    if (flowConfiguration.getLandmarks() == null) {
+      throw new LandmarkNotSetException(
+          "The SessionContinuityInterceptor is enabled, but no 'landmarks' section has been created in the application's form " +
```
Comment (bseeger) [inline src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:65]: Updated message here.

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 9]
```diff
@@ -40,61 +41,63 @@ public SessionContinuityInterceptor(List<FlowConfiguration> flowConfigurations)
...
+              "flows configuration file.");
+    }
+
+    if (flowConfiguration.getLandmarks().getFirstScreen() == null) {
+      throw new LandmarkNotSetException(
+          "The SessionContinuityInterceptor is enabled, but a 'firstScreen' page has not " +
```
Comment (bseeger) [inline src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:71]: Updated message here.

File: src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java [hunk 10]
```diff
@@ -40,61 +41,63 @@ public SessionContinuityInterceptor(List<FlowConfiguration> flowConfigurations)
...
+              "been identified in the 'landmarks' section in the application's form flows configuration file.");
+    }
+
+    String firstScreen = flowConfiguration.getLandmarks().getFirstScreen();
+
+    if (!flowConfiguration.getFlow().containsKey(firstScreen)) {
+      throw new LandmarkNotSetException(String.format(
+          "The form flows configuration file does not contain a screen with the name '%s'. " +
+              "Please make sure to correctly set the 'firstScreen' in the form flows configuration file 'landmarks' section.",
+          firstScreen));
```
Comment (bseeger) [inline src/main/java/formflow/library/interceptors/SessionContinuityInterceptor.java:81]: Updated message here.

File: src/test/java/formflow/library/interceptors/InterceptorMockMvcTest.java [hunk 11]
```diff
@@ -78,7 +79,10 @@ void shouldErrorIfLandmarkIsNotSet() throws Exception {
         .andExpect(result -> {
           Exception resolvedException = result.getResolvedException();
           assertTrue(resolvedException instanceof LandmarkNotSetException, "Expected RuntimeException to be thrown");
-          assertEquals("You have enabled session continuity interception but have not created a landmark section in your applications flow configuration file.", resolvedException.getMessage());
+
+          assertEquals(
+              "The SessionContinuityInterceptor is enabled, but no 'landmarks' section has been created in the application's form flows configuration file.",
```
Comment (bseeger) [inline src/test/java/formflow/library/interceptors/InterceptorMockMvcTest.java:84]: In the tests I only updated error messages to match changes.
